### ISSUE TITLE
[Bug]: Incompatibilities with OpenTelemetry LLM semantics pending release

### ISSUE THREAD

**Issue by codefromthecrypt** [openlit/openlit#342]
### 🐛 What's Going Wrong?

I work on the OpenTelemetry LLM semantics SIG, and did an evaluation of the SDK based on the following sample code and what the semantics pending release 1.27.0 will define.

The sample code produces one span and five metrics.

The span has several exact matches with the pending semantics, but also numerous areas where data is collected as different attributes, different value types or different model locations (e.g. attribute where defined as an event). That said, many differences feel like linting vs major differences. 

The metrics currently have little in common with what's pending release, either by accident or by intent of collecting different signals or collecting it differently.

### 🕵️ Steps to Reproduce
run the following against an ollama hosted codegemma:2b-code model and look at the collector output.

```python
import os
import openlit
from openai import OpenAI

# Set the service name such that it is different from other experiments
application_name = "openlit-python-ollama"
# Default the SDK endpoint ENV variable to localhost
otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318")
# Don't batch spans, as this is a demo
openlit.init(application_name=application_name, otlp_endpoint=otlp_endpoint, disable_batch=True)

def main():
    ollama_host = os.getenv('OLLAMA_HOST', 'localhost')
    # Use the OpenAI endpoint, not the Ollama API.
    base_url = 'http://' + ollama_host + ':11434/v1'
    client = OpenAI(base_url=base_url, api_key='unused')
    messages = [
      {
        'role': 'user',
        'content': '<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>',
      },
    ]
    chat_completion = client.chat.completions.create(model='codegemma:2b-code', messages=messages)
    print(chat_completion.choices[0].message.content)

if __name__ == "__main__":
    main()

```

### 🎯 What Did You Expect?

Semantic evaluation on [spans](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-spans.md).

compatible:
* kind=client
* attributes['gen_ai.system']='openai'
* attributes['gen_ai.operation.name']='chat'
* attributes['gen_ai.response.id']='chatcmpl-411'
* attributes['gen_ai.request.model']='codegemma:2b-code'

deprecated fields:
* attributes['gen_ai.usage.completion_tokens']=60 (deprecated for gen_ai.usage.output_tokens)
* attributes['gen_ai.usage.prompt_tokens']=24 (deprecated for gen_ai.usage.input_tokens)

incompatible:
* name='openai.chat.completions' (should be 'chat codegemma:2b-code')
* attributes['gen_ai.request.top_p']=1 (should be a double 1.0, not int 1)
* attributes['gen_ai.request.max_tokens']='' (should be an int, not a string)
* attributes['gen_ai.request.temperature']=1 (should be a double 1.0, not int 1)
* attributes['gen_ai.request.presence_penalty']=0 (should be a double 0.0, not int 0)
* attributes['gen_ai.request.frequency_penalty']=0 (should be a double 0.0, not int 0)
* attributes['gen_ai.prompt']='<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>'  (should be 'content' in the event attribute `gen_ai.prompt`)
* attributes['gen_ai.response.finish_reason']='stop' (should be 'gen_ai.response.finish_reasons' and an array)
* attributes['gen_ai.completion']='...'  (should be 'content' in the event attribute `gen_ai.completion`)

defined in other conventions:
* attributes['telemetry.sdk.name']='openlit' (Resource)

not yet defined in the standard:
* attributes['gen_ai.endpoint']='openai.chat.completions'
* attributes['gen_ai.environment']='default'
* attributes['gen_ai.application_name']='openlit-python-ollama'
* attributes['gen_ai.request.user']=''
* attributes['gen_ai.request.seed']=''
* attributes['gen_ai.request.is_stream']=false
* attributes['gen_ai.usage.total_tokens']=84
* attributes['gen_ai.usage.cost']=0

Semantic evaluation on
[metrics](https://github.com/open-telemetry/semantic-conventions/blob/main/docs/gen-ai/gen-ai-metrics.md):

#### 'gen_ai.total.requests' (custom)

This is not yet defined in the spec

#### 'gen_ai.usage.total_tokens' (custom)

This is not yet defined in the spec

#### 'gen_ai.usage.completion_tokens' (custom or incompatible)

Two options:
* It meant to be cumulative and is just custom
* It should be a histogram, then should match 'gen_ai.client.token.usage' with 'output' and matching attributes

#### 'gen_ai.usage.prompt_tokens' (custom or incompatible)

Two options:
* It meant to be cumulative and is just custom
* It should be a histogram, then should match 'gen_ai.client.token.usage' with 'input' and matching attributes

#### 'gen_ai.usage.cost' (custom)

This is not yet defined in the spec

### 📸 Any Screenshots?

Example collector log

```
otel-collector      | 2024-07-24T12:01:00.601Z  info    TracesExporter  {"kind": "exporter", "data_type": "traces", "name": "debug", "resource spans": 1, "spans": 1}
otel-collector      | 2024-07-24T12:01:00.602Z  info    ResourceSpans #0
otel-collector      | Resource SchemaURL: 
otel-collector      | Resource attributes:
otel-collector      |      -> service.name: Str(openlit-python-ollama)
otel-collector      |      -> deployment.environment: Str(default)
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      | ScopeSpans #0
otel-collector      | ScopeSpans SchemaURL: 
otel-collector      | InstrumentationScope openlit.otel.tracing 
otel-collector      | Span #0
otel-collector      |     Trace ID       : 51b590ffb975457834cbd6edec5ad881
otel-collector      |     Parent ID      : 
otel-collector      |     ID             : dc5fa811205a8925
otel-collector      |     Name           : openai.chat.completions
otel-collector      |     Kind           : Client
otel-collector      |     Start time     : 2024-07-24 12:00:57.632786259 +0000 UTC
otel-collector      |     End time       : 2024-07-24 12:01:00.598742222 +0000 UTC
otel-collector      |     Status code    : Ok
otel-collector      |     Status message : 
otel-collector      | Attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.endpoint: Str(openai.chat.completions)
otel-collector      |      -> gen_ai.response.id: Str(chatcmpl-411)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      |      -> gen_ai.request.top_p: Int(1)
otel-collector      |      -> gen_ai.request.max_tokens: Str()
otel-collector      |      -> gen_ai.request.user: Str()
otel-collector      |      -> gen_ai.request.temperature: Int(1)
otel-collector      |      -> gen_ai.request.presence_penalty: Int(0)
otel-collector      |      -> gen_ai.request.frequency_penalty: Int(0)
otel-collector      |      -> gen_ai.request.seed: Str()
otel-collector      |      -> gen_ai.request.is_stream: Bool(false)
otel-collector      |      -> gen_ai.prompt: Str(user: <|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>)
otel-collector      |      -> gen_ai.usage.prompt_tokens: Int(24)
otel-collector      |      -> gen_ai.usage.completion_tokens: Int(60)
otel-collector      |      -> gen_ai.usage.total_tokens: Int(84)
otel-collector      |      -> gen_ai.response.finish_reason: Str(stop)
otel-collector      |      -> gen_ai.usage.cost: Int(0)
otel-collector      |      -> gen_ai.completion: Str(  (0)  (1)  (2)  (3)  (4)  (5)  (6)  (7)  (8)  (9)> def hello_world():
otel-collector      | >>> print("Hello, world!")
otel-collector      | 
otel-collector      | )
otel-collector      |   {"kind": "exporter", "data_type": "traces", "name": "debug"}
otel-collector      | 2024-07-24T12:01:00.605Z  info    MetricsExporter {"kind": "exporter", "data_type": "metrics", "name": "debug", "resource metrics": 1, "metrics": 5, "data points": 5}
otel-collector      | 2024-07-24T12:01:00.605Z  info    ResourceMetrics #0
otel-collector      | Resource SchemaURL: 
otel-collector      | Resource attributes:
otel-collector      |      -> service.name: Str(openlit-python-ollama)
otel-collector      |      -> deployment.environment: Str(default)
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      | ScopeMetrics #0
otel-collector      | ScopeMetrics SchemaURL: 
otel-collector      | InstrumentationScope openlit.otel.metrics 0.1.0
otel-collector      | Metric #0
otel-collector      | Descriptor:
otel-collector      |      -> Name: gen_ai.total.requests
otel-collector      |      -> Description: Number of requests to GenAI
otel-collector      |      -> Unit: 1
otel-collector      |      -> DataType: Sum
otel-collector      |      -> IsMonotonic: true
otel-collector      |      -> AggregationTemporality: Cumulative
otel-collector      | NumberDataPoints #0
otel-collector      | Data point attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      | StartTimestamp: 2024-07-24 12:01:00.598679471 +0000 UTC
otel-collector      | Timestamp: 2024-07-24 12:01:00.603022168 +0000 UTC
otel-collector      | Value: 1
otel-collector      | Metric #1
otel-collector      | Descriptor:
otel-collector      |      -> Name: gen_ai.usage.total_tokens
otel-collector      |      -> Description: Number of total tokens processed.
otel-collector      |      -> Unit: 1
otel-collector      |      -> DataType: Sum
otel-collector      |      -> IsMonotonic: true
otel-collector      |      -> AggregationTemporality: Cumulative
otel-collector      | NumberDataPoints #0
otel-collector      | Data point attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      | StartTimestamp: 2024-07-24 12:01:00.59870018 +0000 UTC
otel-collector      | Timestamp: 2024-07-24 12:01:00.603022168 +0000 UTC
otel-collector      | Value: 84
otel-collector      | Metric #2
otel-collector      | Descriptor:
otel-collector      |      -> Name: gen_ai.usage.completion_tokens
otel-collector      |      -> Description: Number of completion tokens processed.
otel-collector      |      -> Unit: 1
otel-collector      |      -> DataType: Sum
otel-collector      |      -> IsMonotonic: true
otel-collector      |      -> AggregationTemporality: Cumulative
otel-collector      | NumberDataPoints #0
otel-collector      | Data point attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      | StartTimestamp: 2024-07-24 12:01:00.598710138 +0000 UTC
otel-collector      | Timestamp: 2024-07-24 12:01:00.603022168 +0000 UTC
otel-collector      | Value: 60
otel-collector      | Metric #3
otel-collector      | Descriptor:
otel-collector      |      -> Name: gen_ai.usage.prompt_tokens
otel-collector      |      -> Description: Number of prompt tokens processed.
otel-collector      |      -> Unit: 1
otel-collector      |      -> DataType: Sum
otel-collector      |      -> IsMonotonic: true
otel-collector      |      -> AggregationTemporality: Cumulative
otel-collector      | NumberDataPoints #0
otel-collector      | Data point attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      | StartTimestamp: 2024-07-24 12:01:00.598714888 +0000 UTC
otel-collector      | Timestamp: 2024-07-24 12:01:00.603022168 +0000 UTC
otel-collector      | Value: 24
otel-collector      | Metric #4
otel-collector      | Descriptor:
otel-collector      |      -> Name: gen_ai.usage.cost
otel-collector      |      -> Description: The distribution of GenAI request costs.
otel-collector      |      -> Unit: USD
otel-collector      |      -> DataType: Histogram
otel-collector      |      -> AggregationTemporality: Cumulative
otel-collector      | HistogramDataPoints #0
otel-collector      | Data point attributes:
otel-collector      |      -> telemetry.sdk.name: Str(openlit)
otel-collector      |      -> gen_ai.application_name: Str(openlit-python-ollama)
otel-collector      |      -> gen_ai.system: Str(openai)
otel-collector      |      -> gen_ai.environment: Str(default)
otel-collector      |      -> gen_ai.operation.name: Str(chat)
otel-collector      |      -> gen_ai.request.model: Str(codegemma:2b-code)
otel-collector      | StartTimestamp: 2024-07-24 12:01:00.598719347 +0000 UTC
otel-collector      | Timestamp: 2024-07-24 12:01:00.603022168 +0000 UTC
otel-collector      | Count: 1
otel-collector      | Sum: 0.000000
otel-collector      | Min: 0.000000
otel-collector      | Max: 0.000000
otel-collector      | ExplicitBounds #0: 0.000000
otel-collector      | ExplicitBounds #1: 5.000000
otel-collector      | ExplicitBounds #2: 10.000000
otel-collector      | ExplicitBounds #3: 25.000000
otel-collector      | ExplicitBounds #4: 50.000000
otel-collector      | ExplicitBounds #5: 75.000000
otel-collector      | ExplicitBounds #6: 100.000000
otel-collector      | ExplicitBounds #7: 250.000000
otel-collector      | ExplicitBounds #8: 500.000000
otel-collector      | ExplicitBounds #9: 750.000000
otel-collector      | ExplicitBounds #10: 1000.000000
otel-collector      | ExplicitBounds #11: 2500.000000
otel-collector      | ExplicitBounds #12: 5000.000000
otel-collector      | ExplicitBounds #13: 7500.000000
otel-collector      | ExplicitBounds #14: 10000.000000
otel-collector      | Buckets #0, Count: 1
otel-collector      | Buckets #1, Count: 0
otel-collector      | Buckets #2, Count: 0
otel-collector      | Buckets #3, Count: 0
otel-collector      | Buckets #4, Count: 0
otel-collector      | Buckets #5, Count: 0
otel-collector      | Buckets #6, Count: 0
otel-collector      | Buckets #7, Count: 0
otel-collector      | Buckets #8, Count: 0
otel-collector      | Buckets #9, Count: 0
otel-collector      | Buckets #10, Count: 0
otel-collector      | Buckets #11, Count: 0
otel-collector      | Buckets #12, Count: 0
otel-collector      | Buckets #13, Count: 0
otel-collector      | Buckets #14, Count: 0
otel-collector      | Buckets #15, Count: 0
otel-collector      |   {"kind": "exporter", "data_type": "metrics", "name": "debug"}
```

### 💻 Your Setup
- OpenLIT Version: [e.g., 0.0.1]
- OpenLIT SDK Version: [e.g., 0.0.3]
- Deployment Method: [Helm, Docker]

### 📝 Additional Notes
>  Got more to say? Tell us here.

### 👐 Want to Help Make It Happen?
- [ ] Yes, I'd like to volunteer and help out with this!

**Comment by amanagarwal042** [comment]
Hey @codefromthecrypt 

> name='openai.chat.completions' (should be 'chat codegemma:2b-code')

Wont be changing this, It doesn't make sense to have the way we have in the conventions and it should be the endpoint that is being hit which is the general norm.

> attributes['gen_ai.request.max_tokens']='' (should be an int, not a string)

It is an int, since its empty we pass an empty string as passing any other default integer would be wrong

Fo the Metrics of Token (prompt and Completion) We have heard its better to have the granuality so we would keep them as seperate metric, hence custom to us. Though we will mention this in the semconv meeting


rest of the issues I am curently working on

**Comment by codefromthecrypt** [comment]
@patcher9 you mind putting a comment about the name stuff here? One of the good outcomes I was hoping from this work is folks approving specs to know the impact to implementors. Maybe this can lead to more careful consideration or active pinging in the future https://github.com/open-telemetry/semantic-conventions/pull/1202

**Comment by codefromthecrypt** [comment]
> It is an int, since its empty we pass an empty string as passing any other default integer would be wrong

maybe there is something to clear up on the semantics about this, too. For example, to omit a value if nil has significance vs whatever the data type's zero value is. This is the first time I've seen a datatype change conditional on nil or not, so maybe someone else with more experience has more commentary about python norms in otel.

@xrmx any thoughts?

**Comment by xrmx** [comment]
> > It is an int, since its empty we pass an empty string as passing any other default integer would be wrong
> 
> maybe there is something to clear up on the semantics about this, too. For example, to omit a value if nil has significance vs whatever the data type's zero value is. This is the first time I've seen a datatype change conditional on nil or not, so maybe someone else with more experience has more commentary about python norms in otel.

Since this is `recommended` maybe it's simpler to don't send it if you don't have it?

**Comment by amanagarwal042** [comment]
> Since this is recommended maybe it's simpler to don't send it if you don't have it?

I can try to set a conditional if the value is not none but ideally wanna avoid adding conditionals as it adds to the latency.

**Comment by codefromthecrypt** [comment]
> I can try to set a conditional if the value is not none but ideally wanna avoid adding conditionals as it adds to the latency.

It is great to be perf centric on instrumentation as yeah we can add a lot of unnecessary overhead sometimes. This is indeed important when inline processing in the request or response path. I totally get it and spent years optimizing clocks access, field mutability, etc. One thing I noticed about this is that sometimes things aren't as bad as they seem, or non-intuitive. Like reading a clock, something much more expensive than field access, is cached so really less than you might guess. There's a role to play in microbenchmarking instrumentation and I even think this is something I might be interested in helping with here at least openai level one day. That said, I think we can fairly assume checking nil is more about clutter of the code than the perf hit. If I'm wrong, you can use this comment and blame me publicly later ;) I'm happy to be wrong sometimes, as that's how I learn too.

**Comment by amanagarwal042** [comment]
Gotcha @codefromthecrypt

I can try that :)
Thanks!

**Comment by codefromthecrypt** [comment]
Did a quick pass here, span name same as before per your notes, and a few incompatible things remain on span:

attributes['gen_ai.prompt']='<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>' (should be 'content' in the event attribute gen_ai.prompt)
attributes['gen_ai.response.finish_reasons']='stop' (should be in an array)
attributes['gen_ai.completion']='...' (should be 'content' in the event attribute gen_ai.completion)

**Comment by amanagarwal042** [comment]
> attributes['gen_ai.prompt']='<|fim_prefix|>def hello_world():<|fim_suffix|><|fim_middle|>' (should be 'content' in the event attribute gen_ai.prompt)
> attributes['gen_ai.response.finish_reasons']='stop' (should be in an array)
> attributes['gen_ai.completion']='...' (should be 'content' in the event attribute gen_ai.completion)

Should be fixed in the PR #354

**Comment by codefromthecrypt** [comment]
sweet when you say there's a package published I'll roll the lint brush again!

**Comment by amanagarwal042** [comment]
Released in `1.17.0` SDK version. 

Not for Prompts and Completions I havent used the format 

```
[{'role': 'user', 'content': 'What is the capital of France?'}]
```

as its captured as a string itself and did not make sense to me why I would want to caupture my content in this format (It would make showing specific prompt content and response in visualizations and analytics very tough. 

We will also refer this issue in the GenAI SIG

**Comment by codefromthecrypt** [comment]
> We will also refer this issue in the GenAI SIG

thanks for offering to contribute upstream so that the SIG results are practical!
